File: DSA/Aanidhay-Aggarwal/Day-18/Day18_q2.java
import java.util.*;

class TreeNode {
    char val;
    int frequency;
    TreeNode left;
    TreeNode right;
    TreeNode(char x) { val = x; frequency = 0; }
}

public class Main {
    public static String decode(TreeNode root, String binaryString) {
        if (root == null) {
            return "";
        }

        StringBuilder result = new StringBuilder();
        TreeNode current = root;

        for (char c : binaryString.toCharArray()) {
            if (c == '0') {
                current = current.left;
            } else {
                current = current.right;
            }

            if (current.left == null && current.right == null) {
                result.append(current.val);
                current = root;
            }
        }

        return result.toString();
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("Enter the number of nodes in the Huffman tree:");
        int n = scanner.nextInt();

        System.out.println("Enter the characters and their frequencies:");
        Map<Character, Integer> frequencyMap = new HashMap<>();
        for (int i = 0; i < n; i++) {
            char c = scanner.next().charAt(0);
            int frequency = scanner.nextInt();
            frequencyMap.put(c, frequency);
        }

        System.out.println("Enter the binary string:");
        String binaryString = scanner.next();

        TreeNode root = createHuffmanTree(frequencyMap);

        String result = decode(root, binaryString);
        System.out.println("Decoded text: " + result);
    }

    public static TreeNode createHuffmanTree(Map<Character, Integer> frequencyMap) {
        PriorityQueue<TreeNode> queue = new PriorityQueue<>((a, b) -> a.frequency - b.frequency);

        for (Map.Entry<Character, Integer> entry : frequencyMap.entrySet()) {
            TreeNode node = new TreeNode(entry.getKey());
            node.frequency = entry.getValue();
            queue.add(node);
        }

        while (queue.size() > 1) {
            TreeNode left = queue.poll();
            TreeNode right = queue.poll();
            TreeNode internalNode = new TreeNode(' ');
            internalNode.left = left;
            internalNode.right = right;
            internalNode.frequency = left.frequency + right.frequency;
            queue.add(internalNode);
        }

        return queue.poll();
    }
}

File: DSA/saumya/DAY18_Q2.py
class HuffmanNode:
    def __init__(self, character, frequency):
        self.character = character
        self.frequency = frequency
        self.left = None
        self.right = None

class decode_text:
    def __init__(self, root):
        self.root = root

    def decode(self, binary_string):
        decoded_text = []
        current_node = self.root
        for bit in binary_string:
            if bit == '0':
                current_node = current_node.left
            else:
                current_node = current_node.right
            if current_node.left is None and current_node.right is None:
                decoded_text.append(current_node.character)
                current_node = self.root
        return ''.join(decoded_text)

node_a = HuffmanNode('A', 3)
node_b = HuffmanNode('B', 1)
node_c = HuffmanNode('C', 1)
internal_node = HuffmanNode('*', 2)
internal_node.left = node_b
internal_node.right = node_c

root = HuffmanNode('*', 5)
root.left = node_a
root.right = internal_node
huffman_tree = decode_text(root)
binary_string = "1001011"
decoded_text = huffman_tree.decode(binary_string)
print("Decoded text:", decoded_text)

